hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/fsdataset/impl/BlockPoolSlice.java

```diff
@@ -1152,4 +1153,14 @@ void setDeleteDuplicateReplicasForTests(
     this.deleteDuplicateReplicas = deleteDuplicateReplicasForTests;
   }
 
+  public File hardLinkOneBlock(File src, File srcMeta, Block dstBlock) throws IOException {
+    File dstMeta = new File(tmpDir,
+        DatanodeUtil.getMetaName(dstBlock.getBlockName(), dstBlock.getGenerationStamp()));
+    HardLink.createHardLink(srcMeta, dstMeta);
+
+    File dstBlockFile = new File(tmpDir, dstBlock.getBlockName());
+    HardLink.createHardLink(src, dstBlockFile);
```

Comment:
Hi @LiuGuH @haiyang1987 , do you use this feature on a large scale? Does creating a lot of hard links have a maintenance impact on the system? Have you considered implementing it based on rename?

Reply:
>>do you use this feature on a large scale? Does creating a lot of hard links have a maintenance impact on the system? 

Now Hadoop version upgrade depends on  linux hardlink.   And hardlink is lightweight and it only link a new pointer points to the original file. 

>> Have you considered implementing it based on rename?
Now  src and dst have different  block name.  And it points the same file address.  @tomscut